### PR TITLE
Elasticsearch: Add trace to logs functionality

### LINKED ISSUE #56367: Elasticsearch: Add trace to Logs functionality
### Summary

Add trace to logs functionality for ES to allow ES users jump into their logs from tempo traces. 

### Problem statement

Currently, ES is not supported in trace to logs functionality. We want to add logic so ES is supported for this feature. 
 
### Scope

Implementation

### Solution

This is hardcoded for splunk + loki here https://github.com/grafana/grafana/blob/2970c6db375e258a6fbf129c21b7210af248d40c/public/app/features/explore/TraceView/createSpanLink.tsx#L115
It just needs a case for elastic so it creates correct query.

And then in `public/app/core/components/TraceToLogs/TraceToLogsSettings.tsx` elasticsearch should be added.

### Resources
Discussed in https://github.com/grafana/grafana/discussions/51746

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What is this feature?**

In tracing data source (Tempo/Zipkin/Jaeger) config page, you can now choose ES data sources in Trace to Logs settings and allow ES users jump into their logs from traces. 

**Why do we need this feature?**

Here is a blog post on why is it valuable (the blog post talks about Splunk, but the same applies here)
https://grafana.com/blog/2022/05/09/new-in-grafana-8.5-how-to-jump-from-traces-to-splunk-logs/

**Who is this feature for?**

ES users

**Which issue(s) does this PR fix?**:

Fixes: https://github.com/grafana/grafana/issues/56367

**Special notes for your reviewer**:

I have added some typescript fixes in files I touched. 

How to test this:

1. Run `make devenv sources=elastic` 
2. Create a new `Tempo` data source (you don't need to run docker block for it, as we you can upload attached trace json for testing):
  - **URL**: http://localhost:3200
  - **Trace to Logs config**
    - **Data source**: `gdev-elasticsearch`
    - **Tags**: `level`, `label`
  When testing, feel free to turn on map tag names, change span time shift, add filter by Span/Trace ID
3. Go to Explore,  choose `Tempo` data source and update following trace:
```
{"batches":[{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"lb"}},{"key":"label","value":{"stringValue":"val1"}},{"key":"opencensus.exporterversion","value":{"stringValue":"Jaeger-Go-2.22.1"}},{"key":"host.name","value":{"stringValue":"07345ebe1f68"}},{"key":"ip","value":{"stringValue":"172.26.0.8"}},{"key":"client-uuid","value":{"stringValue":"57db56d6431793e7"}}],"droppedAttributesCount":0},"instrumentationLibrarySpans":[{"spans":[{"traceId":"AAAAAAAAAAA4WCOI8lqm6w==","spanId":"OFgjiPJapus=","traceState":"","parentSpanId":"","name":"HTTP Client","startTimeUnixNano":1667393901991980000,"endTimeUnixNano":1667393902009652200,"attributes":[{"key":"sampler.type","value":{"stringValue":"const"}},{"key":"sampler.param","value":{"boolValue":true}},{"key":"status.code","value":{"intValue":0}}],"droppedAttributesCount":0,"droppedEventsCount":0,"droppedLinksCount":0}]}]},{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"lb"}},{"key":"label","value":{"stringValue":"val2"}},{"key":"opencensus.exporterversion","value":{"stringValue":"Jaeger-Go-2.22.1"}},{"key":"host.name","value":{"stringValue":"07345ebe1f68"}},{"key":"ip","value":{"stringValue":"172.26.0.8"}},{"key":"client-uuid","value":{"stringValue":"57db56d6431793e7"}}],"droppedAttributesCount":0},"instrumentationLibrarySpans":[{"spans":[{"traceId":"AAAAAAAAAAA4WCOI8lqm6w==","spanId":"NmvvXsnvRyk=","traceState":"","parentSpanId":"OFgjiPJapus=","name":"HTTP GET","kind":"SPAN_KIND_CLIENT","startTimeUnixNano":1667393901992025000,"endTimeUnixNano":1667393902009986000,"attributes":[{"key":"http.status_code","value":{"intValue":200}},{"key":"level","value":{"stringValue":"info"}},{"key":"http.method","value":{"stringValue":"GET"}},{"key":"http.url","value":{"stringValue":"app:80"}},{"key":"net/http.reused","value":{"boolValue":false}},{"key":"net/http.was_idle","value":{"boolValue":false}},{"key":"status.code","value":{"intValue":0}}],"droppedAttributesCount":0,"droppedEventsCount":0,"droppedLinksCount":0,"events":[{"timeUnixNano":1667393901992144000,"attributes":[{"key":"event","value":{"stringValue":"GetConn"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901992423000,"attributes":[{"key":"event","value":{"stringValue":"DNSStart"}},{"key":"host","value":{"stringValue":"app"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901995194000,"attributes":[{"key":"event","value":{"stringValue":"DNSDone"}},{"key":"addr","value":{"stringValue":"172.26.0.7"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901995220000,"attributes":[{"key":"event","value":{"stringValue":"ConnectStart"}},{"key":"network","value":{"stringValue":"tcp"}},{"key":"addr","value":{"stringValue":"172.26.0.7:80"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901996821000,"attributes":[{"key":"event","value":{"stringValue":"ConnectDone"}},{"key":"network","value":{"stringValue":"tcp"}},{"key":"addr","value":{"stringValue":"172.26.0.7:80"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901996910000,"attributes":[{"key":"event","value":{"stringValue":"GotConn"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901997145000,"attributes":[{"key":"event","value":{"stringValue":"WroteHeaders"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901997153000,"attributes":[{"key":"event","value":{"stringValue":"WroteRequest"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902009212000,"attributes":[{"key":"event","value":{"stringValue":"GotFirstResponseByte"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902009983000,"attributes":[{"key":"event","value":{"stringValue":"ClosedBody"}}],"droppedAttributesCount":0,"name":""}]}]}]},{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"app"}},{"key":"label","value":{"stringValue":"val1"}},{"key":"opencensus.exporterversion","value":{"stringValue":"Jaeger-Go-2.22.1"}},{"key":"host.name","value":{"stringValue":"37c87fe98cbf"}},{"key":"ip","value":{"stringValue":"172.26.0.7"}},{"key":"client-uuid","value":{"stringValue":"7c1d9fd22d8f4ca2"}}],"droppedAttributesCount":0},"instrumentationLibrarySpans":[{"spans":[{"traceId":"AAAAAAAAAAA4WCOI8lqm6w==","spanId":"YpF0z9fu0t8=","traceState":"","parentSpanId":"NmvvXsnvRyk=","name":"HTTP GET - root","kind":"SPAN_KIND_SERVER","startTimeUnixNano":1667393901998089000,"endTimeUnixNano":1667393902012023000,"attributes":[{"key":"http.status_code","value":{"intValue":200}},{"key":"http.method","value":{"stringValue":"GET"}},{"key":"http.url","value":{"stringValue":"/"}},{"key":"level","value":{"stringValue":"info"}},{"key":"status.code","value":{"intValue":0}}],"droppedAttributesCount":0,"droppedEventsCount":0,"droppedLinksCount":0},{"traceId":"AAAAAAAAAAA4WCOI8lqm6w==","spanId":"WBwbh0tgGAo=","traceState":"","parentSpanId":"YpF0z9fu0t8=","name":"HTTP Client","startTimeUnixNano":1667393901998368000,"endTimeUnixNano":1667393902007997700,"attributes":[{"key":"status.code","value":{"intValue":0}}],"droppedAttributesCount":0,"droppedEventsCount":0,"droppedLinksCount":0},{"traceId":"AAAAAAAAAAA4WCOI8lqm6w==","spanId":"KnmteGxavEI=","traceState":"","parentSpanId":"WBwbh0tgGAo=","name":"HTTP GET","kind":"SPAN_KIND_CLIENT","startTimeUnixNano":1667393901998727000,"endTimeUnixNano":1667393902011800800,"attributes":[{"key":"http.status_code","value":{"intValue":200}},{"key":"level","value":{"stringValue":"error"}},{"key":"http.method","value":{"stringValue":"GET"}},{"key":"http.url","value":{"stringValue":"db:80"}},{"key":"net/http.reused","value":{"boolValue":false}},{"key":"net/http.was_idle","value":{"boolValue":false}},{"key":"status.code","value":{"intValue":0}}],"droppedAttributesCount":0,"droppedEventsCount":0,"droppedLinksCount":0,"events":[{"timeUnixNano":1667393901998854000,"attributes":[{"key":"event","value":{"stringValue":"GetConn"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393901999076000,"attributes":[{"key":"event","value":{"stringValue":"DNSStart"}},{"key":"host","value":{"stringValue":"db"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902002314000,"attributes":[{"key":"event","value":{"stringValue":"DNSDone"}},{"key":"addr","value":{"stringValue":"172.26.0.2"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902002353000,"attributes":[{"key":"event","value":{"stringValue":"ConnectStart"}},{"key":"network","value":{"stringValue":"tcp"}},{"key":"addr","value":{"stringValue":"172.26.0.2:80"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902003171000,"attributes":[{"key":"event","value":{"stringValue":"ConnectDone"}},{"key":"network","value":{"stringValue":"tcp"}},{"key":"addr","value":{"stringValue":"172.26.0.2:80"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902003271000,"attributes":[{"key":"event","value":{"stringValue":"GotConn"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902003562000,"attributes":[{"key":"event","value":{"stringValue":"WroteHeaders"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902003568000,"attributes":[{"key":"event","value":{"stringValue":"WroteRequest"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902007739000,"attributes":[{"key":"event","value":{"stringValue":"GotFirstResponseByte"}}],"droppedAttributesCount":0,"name":""},{"timeUnixNano":1667393902011799000,"attributes":[{"key":"event","value":{"stringValue":"ClosedBody"}}],"droppedAttributesCount":0,"name":""}]}]}]},{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"db"}},{"key":"label","value":{"stringValue":"val2"}},{"key":"opencensus.exporterversion","value":{"stringValue":"Jaeger-Go-2.22.1"}},{"key":"host.name","value":{"stringValue":"5c5da9b12a0e"}},{"key":"ip","value":{"stringValue":"172.26.0.2"}},{"key":"client-uuid","value":{"stringValue":"22427ee2797437eb"}}],"droppedAttributesCount":0},"instrumentationLibrarySpans":[{"spans":[{"traceId":"AAAAAAAAAAA4WCOI8lqm6w==","spanId":"HREUI7mSYvY=","traceState":"","parentSpanId":"KnmteGxavEI=","name":"HTTP GET - root","kind":"SPAN_KIND_SERVER","startTimeUnixNano":1667393902005430000,"endTimeUnixNano":1667393902006187000,"attributes":[{"key":"http.status_code","value":{"intValue":200}},{"key":"http.method","value":{"stringValue":"GET"}},{"key":"http.url","value":{"stringValue":"/"}},{"key":"level","value":{"stringValue":"error"}},{"key":"status.code","value":{"intValue":0}}],"droppedAttributesCount":0,"droppedEventsCount":0,"droppedLinksCount":0}]}]}]}
```
4.  Then click on traces to logs button.
image

5. Observe that query is correct and results are displayed. You might need to update timerange to last 4 hours/last day (depending on when this is reviewed). In many cases there might not be results - depending on added spanid/traceid filter, timeshift, name mappings. The goal is to make sure query is correct.

6. The produced query is `"spanIDFilter" AND "traceIDFilter" AND label:"val1" AND key:"value2"` . Here is more on lucene query syntax: https://www.elastic.co/guide/en/kibana/master/lucene-query.html. And here is how do we produce it in `importFromAbstractQuery` https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/elasticsearch/LanguageProvider.ts#L21

### CODE REVIEW

**Review by gabor — COMMENTED** [review]
code is fine (two minor issues, left comments), and it is working correctly.. well, as far as i can test it (i'm not good at tracing) 👍

**Review by gabor — APPROVED** [review]
LGTM!

File: public/app/features/explore/TraceView/createSpanLink.test.ts [hunk 1]
```diff
@@ -589,6 +593,154 @@ describe('createSpanLinkFactory', () => {
       );
     });
   });
+
+  describe('elasticsearch link', () => {
+    const elasticsearchUID = 'elasticsearchUID';
+
+    beforeAll(() => {
+      setDataSourceSrv({
+        getInstanceSettings(uid: string) {
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply public/app/features/explore/TraceView/createSpanLink.test.ts]: I fixed everywhere but here 🙈 . It is not needed, so I removed it in https://github.com/grafana/grafana/pull/58063/commits/747d5d5c66bfd8e9b5e826627507b902ac7a3b71

File: public/app/features/explore/TraceView/createSpanLink.test.ts [hunk 2]
```diff
@@ -589,6 +593,154 @@ describe('createSpanLinkFactory', () => {
...
+      const linkDef = links?.logLinks?.[0];
+      expect(linkDef).toBeDefined();
+      expect(linkDef!.href).toContain(
+        encodeURIComponent(
+          `datasource":"${elasticsearchUID}","queries":[{"query":"cluster:\\"cluster1\\" AND hostname:\\"hostname1\\"","refId":"","metrics":[{"id":"1","type":"logs"}]}]`
+        )
+      );
+    });
+
+    it('automatically timeshifts the time range by one second in a splunk query', () => {
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply public/app/features/explore/TraceView/createSpanLink.test.ts]: Fixed in https://github.com/grafana/grafana/pull/58063/commits/747d5d5c66bfd8e9b5e826627507b902ac7a3b71

**Review by aocenas — APPROVED** [review]

**Pull request merged**